kernel/terminal.c

```diff
@@ -28,83 +31,127 @@ enum vga_color {
+    uint16_t empty_color = VGA_COLOR_BLACK | VGA_COLOR_GREEN << 4;
+    for(size_t y=0;y<VGA_HEIGHT;++y)
+        for(size_t x=0;x<VGA_WIDTH;++x){
+			size_t index = y * VGA_WIDTH + x;
+            terminal_buffer[index] = (uint16_t) ((unsigned char)BLANK_CHAR) | (uint16_t) empty_color << 8;
+        }
 }
- 
-uint16_t vga_entry(unsigned char uc, uint8_t color) {
-	return (uint16_t) uc | (uint16_t) color << 8;
+
+/* Initialise a state, i.e. put every char on blank and the default color everywhere */
+void terminal_initialize_state(terminal_state* state){
+    state->position=0;
+    state->bg_color = VGA_COLOR_BLACK;
+    state->fg_color = VGA_COLOR_WHITE;
+    for(size_t y=0;y<VGA_HEIGHT;++y)
+        for(size_t x=0;x<VGA_WIDTH;++x){
+			size_t index = y * VGA_WIDTH + x;
+            state->entries[index].value = BLANK_CHAR;
+        }
+    __terminal_clear(state);
 }
- 
-size_t strlen(const char* str) {
-	size_t len = 0;
-	while (str[len])
-		len++;
-	return len;
+
+/* Copy the state from one to the other */
+void terminal_state_copy(terminal_state* from, terminal_state* to){
+    to->position=from->position;
+    to->bg_color = from->bg_color;
+    to->fg_color = from->fg_color;
+    for(size_t y=0;y<VGA_HEIGHT;++y)
+        for(size_t x=0;x<VGA_WIDTH;++x){
+			size_t index = y * VGA_WIDTH + x;
+            to->entries[index].value = from->entries[index].value;
+        }
 }
 
-void terminal_initialize(void) {
-	terminal_row = 0;
-	terminal_column = 0;
-	terminal_color = vga_entry_color(VGA_COLOR_LIGHT_GREY, VGA_COLOR_BLACK);
-	terminal_buffer = (uint16_t*) 0xB8000;
-	for (size_t y = 0; y < VGA_HEIGHT; y++) {
-		for (size_t x = 0; x < VGA_WIDTH; x++) {
-			const size_t index = y * VGA_WIDTH + x;
-			terminal_buffer[index] = vga_entry(' ', terminal_color);
-		}
-	}
+/* This function will update the terminal buffer with the given state */
+void terminal_update(terminal_state *state){
+    __terminal_clear(state);
+    uint16_t* terminal_buffer = (uint16_t*) TERMINAL_BUFFER_START;
+    for(size_t y=0;y<VGA_HEIGHT;++y)
+        for(size_t x=0;x<VGA_WIDTH;++x){
+			size_t index = y * VGA_WIDTH + x;
+            terminal_entry entry = state->entries[index];
+            uint8_t color = entry.fg_color | entry.bg_color << 4;
+            terminal_buffer[index] = (uint16_t) ((unsigned char)entry.value) | (uint16_t) color << 8;
+        }
 }
- 
-void terminal_setcolor(uint8_t color) {
-	terminal_color = color;
+
+/* Update the background color */
+void __terminal_set_bg_color(terminal_state* state, enum vga_color color){
+    state->bg_color = color;
 }
- 
-void terminal_putentryat(char c, uint8_t color, size_t x, size_t y) {
-	const size_t index = y * VGA_WIDTH + x;
-	terminal_buffer[index] = vga_entry(c, color);
+
+/* Update the foreground color */
+void __terminal_set_fg_color(terminal_state* state, enum vga_color color){
+    state->fg_color = color;
 }
- 
-void terminal_putchar(char c) {
```

Comment:
We want to keep a `terminal_putchar` publicly available, but one that does update the screen when it is done.